src/WebJobs.Extensions.DurableTask/Listener/DurableTaskMetricsProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if !FUNCTIONS_V1
+using System;
+using System.Threading.Tasks;
+using DurableTask.AzureStorage.Monitoring;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Listener
+{
+    internal class DurableTaskMetricsProvider
+    {
+        private readonly string functionName;
+        private readonly string hubName;
+        private readonly ILogger logger;
+        private readonly CloudStorageAccount storageAccount;
+
+        private DisconnectedPerformanceMonitor performanceMonitor;
+
+        public DurableTaskMetricsProvider(string functionName, string hubName, ILogger logger, DisconnectedPerformanceMonitor performanceMonitor, CloudStorageAccount storageAccount)
+        {
+            this.functionName = functionName;
+            this.hubName = hubName;
+            this.logger = logger;
+            this.performanceMonitor = performanceMonitor;
+            this.storageAccount = storageAccount;
+        }
+
+        public virtual async Task<DurableTaskTriggerMetrics> GetMetricsAsync()
+        {
+            DurableTaskTriggerMetrics metrics = new DurableTaskTriggerMetrics();
+
+            // Durable stores its own metrics, so we just collect them here
+            PerformanceHeartbeat heartbeat = null;
+            try
+            {
+                DisconnectedPerformanceMonitor performanceMonitor = this.GetPerformanceMonitor();
+                heartbeat = await performanceMonitor.PulseAsync();
+            }
+            catch (StorageException e)
+            {
+                this.logger.LogWarning("{details}. Function: {functionName}. HubName: {hubName}.", e.ToString(), this.functionName, this.hubName);
+            }
+
+            if (heartbeat != null)
+            {
+                metrics.PartitionCount = heartbeat.PartitionCount;
+                metrics.ControlQueueLengths = JsonConvert.SerializeObject(heartbeat.ControlQueueLengths);
+                metrics.ControlQueueLatencies = JsonConvert.SerializeObject(heartbeat.ControlQueueLatencies);
+                metrics.WorkItemQueueLength = heartbeat.WorkItemQueueLength;
+                if (heartbeat.WorkItemQueueLatency > TimeSpan.Zero)
+                {
+                    metrics.WorkItemQueueLatency = heartbeat.WorkItemQueueLatency.ToString();
+                }
+            }
+
+            return metrics;
+        }
+
+        private DisconnectedPerformanceMonitor GetPerformanceMonitor()
+        {
+            if (this.performanceMonitor == null)
```

Comment:
In what situations do we expect `this.performanceMonitor` to be null?

Reply:
I'm not sure, this logic predates me. You can find it already in our repo here: https://github.com/Azure/azure-functions-durable-extension/blob/5c6172ca353b5ce59a301170fb3daeb57ae4f4f3/src/WebJobs.Extensions.DurableTask/Listener/DurableTaskScaleMonitor.cs#L53-L65

In fact, I did not meant to duplicate this logic (it is now here and in the `DurableTaskScaleMonitor` class), I must have forgotten to put it on a shared location

Reply:
Let me try to refactor this so that we're not duplicating this logic. Looking into that shortly.

Reply:
duplication removed in this commit: https://github.com/Azure/azure-functions-durable-extension/pull/2452/commits/2e6a3175c4c6a7e229aad58b4aed3ff053f76700